# Dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website

The aim of this project is to clean the data and analyze the **included used car listings** in dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. 

Used cars database download [kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data)


## The data dictionary provided with data is as follows:
>**dateCrawled** - When this ad was first crawled. All field-values are taken from this date.

>**name** - Name of the car.

>**seller** - Whether the seller is private or a dealer.

>**offerType** - The type of listing

>**price** - The price on the ad to sell the car.

>**abtest** - Whether the listing is included in an A/B test.

>**vehicleType** - The vehicle Type.

>**yearOfRegistration** - The year in which the car was first registered.

>**gearbox** - The transmission type.

>**powerPS** - The power of the car in PS.

>**model** - The car model name.

>**kilometer** - How many kilometers the car has driven.

>**monthOfRegistration** - The month in which the car was first registered.

>**fuelType** - What type of fuel the car uses.

>**brand** - The brand of the car.

>**notRepairedDamage** - If the car has a damage which is not yet repaired.

>**dateCreated** - The date on which the eBay listing was created.

>**nrOfPictures** - The number of pictures in the ad.

>**postalCode** - The postal code for the location of the vehicle.

>**lastSeenOnline** - When the crawler saw this ad last online.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
autos = pd.read_csv("/content/drive/My Drive/Data Analyst In Python/Modulo3: Introduction to NumPy & Pandas/proyecto modulo3/autos.csv",encoding="Latin-1")

# **Information about Dataframe**

In [ ]:
print("Number of rows: ",autos.shape[0])
print("Number of columns: ",autos.shape[1])

Number of rows:  50000
Number of columns:  20


**First 5 rows and last 5 columns**

In [ ]:
autos.head()

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

In [ ]:
autos.tail()

dateCrawled  ...             lastSeen
49995  2016-03-27 14:38:19  ...  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  ...  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  ...  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  ...  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  ...  2016-04-06 21:18:48

[5 rows x 20 columns]

**Dataframe info**

In [ ]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

About the dataframe shape **expect that each column has at most 5,000 datas**. Taking into account the above **using dataframe.info() method ** can be conclude that columns **vehicleType, gearbox, model, fuelType** and **notRepairedDamage** contain null data.

The dataset contains 20 columns, **most of which are strings.**

For analize columns **price** and **odometer (Kilometers driven)** is necesary delete characters "$$" and "km" taking into account its possible variations, for example: kms, Kms, KM, $ xxx, xxx$, etc. 

Column **name** use camelcase and can give information about cars model offered but is necessary fix each name maybe capitalizing or using  snakecase for avoid leave data unused

Finally could be adequate **modificate some labels** for more clarity and comprehension its information

# Data Cleaning

Rewording some columns names based to be more descriptive.

In [ ]:
print(autos.columns)

reword = { "yearOfRegistration" : "registration_year",
           "monthOfRegistration" : "registration_month",
           "notRepairedDamage" : "unrepaired_damage",
           "dateCreated" : "ad_created",
           "powerPS" : "power_PS",
           "odometer" : "odometer_km"
         }

def change_case(str): 
    res = [str[0].lower()] 
    for c in str[1:]: 
        if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'): 
            res.append('_') 
            res.append(c.lower()) 
        else: 
            res.append(c) 
      
    return ''.join(res)

for label in autos.columns:
    if label not in reword:
        reword[label] = change_case(label)

autos.rename(reword, axis=1, inplace=True)
autos.head(3)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37

[3 rows x 20 columns]

A dictionary called **reword** is created in order to rename some columns with a specific name. Then from a function called **[change_case()](https://www.geeksforgeeks.org/python-program-to-convert-camel-case-string-to-snake-case/ )** proceed to modify the rest of the columns in camelcase to snakecase format, the result of each modification is stored in the ** reword ** dictionary. Finally the columns are renamed using the same dictionary and the method ** dataframe.rename( )**.

**Verification that columns "price" and "odometer_km" are not numeric values**

In [ ]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_PS               int64
model                 object
odometer_km           object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

**price, power_PS and odometer_km columns: Text to numeric values**

In [ ]:
# Characters "$" and "km" are delete for convert its respective columns
# to numeric values

autos["price"] = autos["price"].str.replace("$","").str.replace(",","")
autos["price"] = autos["price"].astype(float)

autos["odometer_km"] = autos["odometer_km"].str.replace("km","").str.replace(",","")
autos["odometer_km"] = autos["odometer_km"].astype(float)

autos["power_PS"] = autos["power_PS"].astype(float)

In [ ]:
print("dtype column price: ",autos["price"].dtypes)
print("dtype column odometer_km: ",autos["odometer_km"].dtypes)
print("dtype column power_PS: ",autos["power_PS"].dtypes)

dtype column price:  float64
dtype column odometer_km:  float64
dtype column power_PS:  float64


**Exploring non-numeric categories**

In [ ]:
autos.describe(include=['O'])

date_crawled  ...            last_seen
count                 50000  ...                50000
unique                48213  ...                39481
top     2016-03-23 18:39:34  ...  2016-04-07 06:17:27
freq                      3  ...                    8

[4 rows x 13 columns]

**Exploring the numerical categories**

In [ ]:
autos.describe()

price  registration_year  ...  nr_of_pictures   postal_code
count  5.000000e+04       50000.000000  ...         50000.0  50000.000000
mean   9.840044e+03        2005.073280  ...             0.0  50813.627300
std    4.811044e+05         105.712813  ...             0.0  25779.747957
min    0.000000e+00        1000.000000  ...             0.0   1067.000000
25%    1.100000e+03        1999.000000  ...             0.0  30451.000000
50%    2.950000e+03        2003.000000  ...             0.0  49577.000000
75%    7.200000e+03        2008.000000  ...             0.0  71540.000000
max    1.000000e+08        9999.000000  ...             0.0  99998.000000

[8 rows x 7 columns]

It is also found that the column ** "postal_code" ** should not be treated as a quantitative category because according to [Wikipedia](https://en.wikipedia.org/wiki/Postal_code), the postal code is a scheme that is assigned to different areas or places in a country, a code that serves to facilitate and mechanize the routing of a piece of mail.

Exploring the numerical categories it can be seen that all the values of the column ** "nr_of_pictures" ** are zero.

In the column **"price"** find as minimum value 0 and maximum $100 Millions. This requires inspection

Column **"registration_year"** find as minimum year 1,000 and maximum 9,999. This requires inspection

Column **"power PS"** find as minimum power is 0.0 PS. This requires inspection

Column **"registration_month"**  find as minimum month 0 and maximum 12. This requires inspection

It is still pending to decide what to do with the null values of the columns ** vehicleType, gearbox, model, fuelType ** and ** notRepairedDamage **, but most likely it will be kept because, despite not presenting all the information, they can be used to carry out analysis regarding the brand and dates.

Finally, to finish cleaning the data, it would be necessary to convert german words to its english counterparts to easily the analizis

**postal_code columns: numeric values to text**

In [ ]:
autos["postal_code"] = autos["postal_code"].astype(str)
print("dtype column postal_code: ",autos["postal_code"].dtypes)

dtype column postal_code:  object


**Column "price":**

About the prices you can see some values that look unrealistically high or low.

About odometer the high or low values looks normal

In [ ]:
print("Top five cars prices:\n")
print(autos["price"].value_counts().sort_index(ascending=False).head(),"\n")
print("Bottom five prices cars:\n")
print(autos["price"].value_counts().sort_index(ascending=True).head(),"\n")

print("Top five odometer cars:\n")
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head(),"\n")
print("Bottom five odometer cars:\n")
print(autos["odometer_km"].value_counts().sort_index(ascending=True).head())

Top five cars prices:

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
Name: price, dtype: int64 

Bottom five prices cars:

0.0    1421
1.0     156
2.0       3
3.0       1
5.0       2
Name: price, dtype: int64 

Top five odometer cars:

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64 

Bottom five odometer cars:

5000.0     967
10000.0    264
20000.0    784
30000.0    789
40000.0    819
Name: odometer_km, dtype: int64


Atypical data are found in prices and it is decided to eliminate them so that the average price of used cars is not biased:

**Only the values of the exterior and interior fences will be taken into account as of Q3 because from Q2 it does not make sense to consider values less than or equal to zero**

>*1) Find the interquartile range.*

>*2) Find the "inner fences" for the data set. *

>*3) Find the "outer fences" for the data set. *


In [ ]:
rangeQ = autos["price"].quantile(q=0.75) - autos["price"].quantile(q=0.25)
print("Interquartile range: ",rangeQ)

innerFences = rangeQ * 1.5
inner_q3 = autos["price"].quantile(q=0.75) + innerFences
print("inner Q3: ",inner_q3)

outerFences = rangeQ * 3
outer_q3 = autos["price"].quantile(q=0.75) + outerFences
print("outer Q3: ",outer_q3)

Interquartile range:  6100.0
inner Q3:  16350.0
outer Q3:  25500.0


**5,205 rows were removed taking into account the price criteria between € 1 to € 16,350**

Consider the statistical  descriptivo called the **coefficient of variation**. For this, I took into account the descriptive statistics of the result of selecting rows with prices from 1 to "inner fence of Q3". For this selection of data, the coefficient of variation is approximately 94%. As [coefficient of variation](https://es.wikipedia.org/wiki/Coeficiente_de_variaci%C3%B3n) is a relative interpretation of the degree of variability, regardless of the scale of the variable, I allowed a maximum coefficient of 100%. The CV tells us how representative the arithmetic mean of the data set is.



*Quantitative data:*

In [ ]:
autos_selected = autos[autos["price"].between(1, 16350)].copy()
autos_selected.describe()

price  registration_year  ...  registration_month  nr_of_pictures
count  44795.000000        44795.00000  ...        44795.000000         44795.0
mean    4089.433620         2003.97339  ...            5.744012             0.0
std     3841.429247           74.81955  ...            3.707064             0.0
min        1.000000         1000.00000  ...            0.000000             0.0
25%     1150.000000         1999.00000  ...            3.000000             0.0
50%     2700.000000         2003.00000  ...            6.000000             0.0
75%     5999.000000         2007.00000  ...            9.000000             0.0
max    16350.000000         9999.00000  ...           12.000000             0.0

[8 rows x 6 columns]

*Qualitative data:*

In [ ]:
autos_selected.describe(include=['O'])

date_crawled         name  ... postal_code            last_seen
count                 44795        44795  ...       44795                44795
unique                43358        34509  ...        6831                36473
top     2016-03-19 17:36:18  Ford_Fiesta  ...       10115  2016-04-07 06:17:27
freq                      3           76  ...         103                    8

[4 rows x 14 columns]

**Null values (vehicle_type, gearbox, model, fuel_type and unrepaired_damage):**

About Columns with null values I proceed to fill these with label "Unknown"

In [ ]:
# autos_selected.loc[autos_selected["vehicle_type"].isnull() == True,"vehicle_type"] = "unkown"
# autos_selected.loc[autos_selected["gearbox"].isnull() == True,"gearbox"] = "unkown"
# autos_selected.loc[autos_selected["model"].isnull() == True,"model"] = "unkown"
# autos_selected.loc[autos_selected["fuel_type"].isnull() == True,"fuel_type"] = "unkown"
# autos_selected.loc[autos_selected["unrepaired_damage"].isnull() == True,"unrepaired_damage"] = "unkown"

#For more efficiency use method dt.fillna("Nan",inplace = True)
autos_selected.fillna("unkown",inplace = True)

In [ ]:
autos_selected.describe(include=['O'])

date_crawled         name  ... postal_code            last_seen
count                 44795        44795  ...       44795                44795
unique                43358        34509  ...        6831                36473
top     2016-03-19 17:36:18  Ford_Fiesta  ...       10115  2016-04-07 06:17:27
freq                      3           76  ...         103                    8

[4 rows x 14 columns]

**Date columns**

This columns with format (**YY-MM-DD**)are identified as string values by pandas:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website

This columns with format (**MM**  and **YY**) are represented as numeric values
- `registration_month`: from the website
- `registration_year`: from the website

The sales ads select to the dataframe were published in the website from october 10th 2015 to april 7th 2016.

This ads were crawler from march to april 2016.

In [ ]:
print(autos_selected["ad_created"].value_counts(normalize=True).sort_index(ascending=True).head(),'\n')
print(autos_selected["ad_created"].value_counts(normalize=True).sort_index(ascending=True).tail())

2015-08-10 00:00:00    0.000022
2015-09-09 00:00:00    0.000022
2015-11-10 00:00:00    0.000022
2015-12-05 00:00:00    0.000022
2015-12-30 00:00:00    0.000022
Name: ad_created, dtype: float64 

2016-04-03 00:00:00    0.038799
2016-04-04 00:00:00    0.036701
2016-04-05 00:00:00    0.011653
2016-04-06 00:00:00    0.003237
2016-04-07 00:00:00    0.001183
Name: ad_created, dtype: float64


In [ ]:
print(autos_selected["date_crawled"].value_counts(normalize=True).sort_index(ascending=True).head(),'\n')
print(autos_selected["date_crawled"].value_counts(normalize=True).sort_index(ascending=True).tail())

2016-03-05 14:06:30    0.000022
2016-03-05 14:06:40    0.000022
2016-03-05 14:07:04    0.000022
2016-03-05 14:07:08    0.000022
2016-03-05 14:07:21    0.000022
Name: date_crawled, dtype: float64 

2016-04-07 14:07:04    0.000022
2016-04-07 14:30:09    0.000022
2016-04-07 14:30:26    0.000022
2016-04-07 14:36:44    0.000022
2016-04-07 14:36:55    0.000022
Name: date_crawled, dtype: float64


In [ ]:
print(autos_selected["last_seen"].value_counts(normalize=True).sort_index(ascending=True).head(),'\n')
print(autos_selected["last_seen"].value_counts(normalize=True).sort_index(ascending=True).tail())

2016-03-05 14:45:46    0.000022
2016-03-05 14:46:02    0.000022
2016-03-05 14:49:34    0.000022
2016-03-05 15:16:11    0.000022
2016-03-05 15:16:47    0.000022
Name: last_seen, dtype: float64 

2016-04-07 14:58:44    0.000067
2016-04-07 14:58:45    0.000022
2016-04-07 14:58:46    0.000022
2016-04-07 14:58:48    0.000067
2016-04-07 14:58:50    0.000067
Name: last_seen, dtype: float64


About the year of car registration, the only values that should make sense are the years 1800 to 2016

- There are Registration year from 2017 to 9999

In [ ]:
autos_selected["registration_year"].value_counts().sort_index(ascending = False).head(12)

9999       2
8888       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     457
2017    1360
2016    1178
Name: registration_year, dtype: int64

- For below 1800 There are Registration year from 1000 to 1111

In [ ]:
autos_selected["registration_year"].value_counts().sort_index(ascending = False).tail(8)

1937    3
1934    2
1929    1
1910    5
1800    2
1111    1
1001    1
1000    1
Name: registration_year, dtype: int64

Dedicated to select the cars with registration from the year 1910 since the previous ones to this present uncertain information.

In [ ]:
autos_selected[["name","registration_year"]].sort_values("registration_year").head(6)

name  registration_year
22316  VW_Kaefer.__Zwei_zum_Preis_von_einem.               1000
49283                             Citroen_HY               1001
24511            Trabant__wartburg__Ostalgie               1111
32585                             UNFAL_Auto               1800
10556                             UNFAL_Auto               1800
45157                             Motorhaube               1910

Rows selected:  42959

Until now, the info removed from the dataframe corresponds to  7041  rows

In [ ]:
autos_selected = autos_selected[autos_selected["registration_year"].between(1910, 2016)]

**Power PS equal to cero and months equal to cero**

- Powers of 0 PS are meaningless for vehicles unless they correspond to vehicles without a motor. I decide to keep the ranges because I think they can still give me general pricing information, however I decide to tag these values with the tag "unknown" upon completion of the cleanup. For now I replace 0 PS with null values to be able to operate on the power_PS column.

- Since the months range from 1 to 12, the months equal to zero I decide to label them as "unknown"

In [ ]:
autos_selected["power_PS"] = autos_selected["power_PS"].replace(0,np.nan)
autos_selected["registration_month"] = autos_selected["registration_month"].replace(0,"unknown")

**Seller and offerType**

- Approximately 99% of used vehicles correspond to private sellers
- 100% of offers correspond to Angebot
- All used cars for sale have no image on their ad

This information does not present considerable variability and for this reason I decide to remove these coolumns since they do not contribute anything significant to the analysis.


In [ ]:
autos_selected["seller"].value_counts(normalize = True) 

privat        0.999977
gewerblich    0.000023
Name: seller, dtype: float64

In [ ]:
autos_selected["offer_type"].value_counts(normalize = True) 

Angebot    1.0
Name: offer_type, dtype: float64

In [ ]:
autos_selected["nr_of_pictures"].value_counts(normalize = True) 

0    1.0
Name: nr_of_pictures, dtype: float64

In [ ]:
autos_selected = autos_selected.drop(["seller","offer_type","nr_of_pictures"],axis = 1)

In [ ]:
autos_selected.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 17 columns]

**Replace German words with English counterparts**

I identify the German words that translated into English can facilitate analysis and proceed to replace

In [ ]:
print(autos_selected["abtest"].unique())
print(autos_selected["vehicle_type"].unique())
print(autos_selected["gearbox"].unique())
print(autos_selected["fuel_type"].unique())
print(autos_selected["unrepaired_damage"].unique())

['control' 'test']
['bus' 'limousine' 'kleinwagen' 'kombi' 'unkown' 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' 'unkown']
['lpg' 'benzin' 'diesel' 'unkown' 'cng' 'hybrid' 'elektro' 'andere']
['nein' 'unkown' 'ja']


In [ ]:
autos_selected = autos_selected.replace(
{
    'bus': 'bus',
    'limousine': 'limousine',
    'kleinwagen': 'small car',
    'kombi': 'station wagon',
    'coupe': 'coupe',
    'suv': 'suv',
    'cabrio': 'convertible',
    'andere': 'others',
    'manuell': 'manual',
    'automatik': 'automatic',
    'benzin': 'gasoline',
    'elektro': 'electric',
    'nein': 'no',
    'unkown': 'unknown',
    'ja': 'yes'
})

In [ ]:
autos_selected["vehicle_type"].head(10)

0              bus
1        limousine
2        limousine
3        small car
4    station wagon
5              bus
6        limousine
7        limousine
8          unknown
9              bus
Name: vehicle_type, dtype: object

# Exploration and analysis

## **Registration year**

The selection is verified with the column "registration_year" in ascending order, showing its first five rows and its last five rows.

In [ ]:
autos_selected[["name","registration_year"]].sort_values("registration_year").head()

name  registration_year
45157                        Motorhaube               1910
30781  Opel_Calibra_V6_DTM_Bausatz_1:24               1910
28693                    Renault_Twingo               1910
22659                      Opel_Corsa_B               1910
3679                         Suche_Auto               1910

In [ ]:
autos_selected[["name","registration_year"]].sort_values("registration_year").tail()

name  registration_year
41972               Ideales_Baustellenfahrzeug               2016
39929     BMW_316_i_compact_blau_99_Tuev_06/16               2016
32471                   Verkaufe_opel_zafira_A               2016
23304               VK_MONDEO_MIT_TÜV_DEZ_2016               2016
25358  Audi_A6_2.7_Quattro_BiTurbo_230_Ps_Tuev               2016

- Used cars for sale with registration year between 1910 and 2016, the highest frequency corresponds to those with registration year 2000 (approximately 7,3%)

- 51% of used cars have their registration years between 1999 and 2006.


In [ ]:
print(autos_selected["registration_year"].value_counts(normalize=True).head(8),'\n')
print(round(autos_selected["registration_year"].value_counts(normalize=True).head(8).sum()*100,1),"%")

2000    0.073140
2005    0.067227
1999    0.067227
2003    0.062129
2004    0.061920
2001    0.060895
2006    0.060127
2002    0.057264
Name: registration_year, dtype: float64 

51.0 %


## **Name and Brand**

Top 5 mean price by brands

In [ ]:
brands = autos_selected["brand"].unique()
mean_brands = {}
mean_mileage = {}
mean_power_PS = {}

for brand in brands:
    row =  autos_selected[autos_selected["brand"] == brand]
    row_mean = row["price"].mean()
    row_mean_mileage = row["odometer_km"].mean()
    row_mean_power = row["power_PS"].mean(skipna = True)
    
    mean_brands[brand] = (round(row_mean,2),round(row_mean_mileage,2),round(row_mean_power,2))
    
mean_price_top_brand = []

    
for k,v in mean_brands.items():
    key_val_as_tuple = (mean_brands[k], k)
    mean_price_top_brand.append(key_val_as_tuple)

mean_price_top_brand = sorted(mean_price_top_brand, reverse = True)

brand_ = []
price_power = []

for brand in mean_price_top_brand:
    brand_.append(brand[1].capitalize())
    price_power.append(brand[0])

top_mean_prices_brands = pd.DataFrame(price_power,index = brand_,columns = ["mean_price","mean_mileage","mean_power_PS"])
  

**Regarding the price**

- The maximum average to minimum average selling price by brand 

In [ ]:
top_mean_prices_brands.head()

mean_price  mean_mileage  mean_power_PS
Porsche        9493.25     115555.56         246.25
Mini           8362.60      98663.66         123.49
Jeep           7172.62     140697.67         184.10
Land_rover     6605.80     141517.86         144.10
Dacia          5915.53      84268.29          89.04

In [ ]:
top_mean_prices_brands.tail()

mean_price  mean_mileage  mean_power_PS
Renault      2281.46     128892.71          94.08
Trabant      1790.86      54538.46          27.65
Daihatsu     1636.20     116410.26          68.48
Rover        1243.31     137459.02         110.20
Daewoo       1049.00     121642.86          80.59

**Regarding the mileage:**
- The highest mileage to the lowest-mileage 

In [ ]:
top_mean_prices_brands.sort_values("mean_mileage",ascending=False).head()

mean_price  mean_mileage  mean_power_PS
Saab           3005.22     145526.32         160.63
Land_rover     6605.80     141517.86         144.10
Volvo          3930.00     140810.47         152.24
Jeep           7172.62     140697.67         184.10
Audi           5703.90     139701.38         162.32

In [ ]:
top_mean_prices_brands.sort_values("mean_mileage",ascending=False).tail()

mean_price  mean_mileage  mean_power_PS
Mini               8362.60      98663.66         123.49
Sonstige_autos     5605.35      94561.64         142.71
Dacia              5915.53      84268.29          89.04
Lada               2688.30      83518.52          76.05
Trabant            1790.86      54538.46          27.65

**Regarding the power PS**

The PS power in the German metric system is the equivalent of the HP power in the English system where **1 PS = 0.9858 HP**

- The highest power PS to the lowest power PS

In [ ]:
top_mean_prices_brands.sort_values("mean_power_PS",ascending=False).head()

mean_price  mean_mileage  mean_power_PS
Porsche     9493.25     115555.56         246.25
Jaguar      5610.72     138018.87         223.83
Jeep        7172.62     140697.67         184.10
Bmw         5649.20     138833.18         169.92
Audi        5703.90     139701.38         162.32

In [ ]:
top_mean_prices_brands.sort_values("mean_power_PS",ascending=False).tail()

mean_price  mean_mileage  mean_power_PS
Fiat         2740.82     117449.66          80.00
Lada         2688.30      83518.52          76.05
Smart        3560.12      99431.82          73.45
Daihatsu     1636.20     116410.26          68.48
Trabant      1790.86      54538.46          27.65

**Most common brand/model combinations**

In [ ]:
brand_model = autos_selected["brand"].str.capitalize() + " " + autos_selected["model"]

In [ ]:
brand_model.value_counts().head(10)

Volkswagen golf           3487
Bmw 3er                   2431
Volkswagen polo           1604
Opel corsa                1589
Opel astra                1339
Volkswagen passat         1299
Audi a4                   1126
Mercedes_benz c_klasse    1005
Bmw 5er                    991
Mercedes_benz e_klasse     813
dtype: int64

In [ ]:
brand_model.value_counts().tail(10)

Lancia kappa                     2
Ford b_max                       1
Bmw i3                           1
Jaguar unknown                   1
Land_rover unknown               1
Audi 200                         1
Land_rover others                1
Volvo v60                        1
Rover discovery                  1
Land_rover range_rover_evoque    1
dtype: int64

**Most common brand/vehicle_type combinations**

In [ ]:
brand_vehicle = autos_selected["brand"].str.capitalize() + " " + autos_selected["vehicle_type"]

In [ ]:
brand_vehicle.value_counts().head(10)

Volkswagen limousine        2472
Volkswagen small car        2422
Bmw limousine               2209
Volkswagen station wagon    1673
Mercedes_benz limousine     1660
Opel small car              1604
Audi station wagon          1330
Audi limousine              1310
Volkswagen bus              1176
Opel limousine              1045
dtype: int64

In [ ]:
brand_vehicle.value_counts().tail(20)

Jaguar unknown              2
Lada small car              1
Smart others                1
Saab suv                    1
Skoda coupe                 1
Land_rover station wagon    1
Jeep convertible            1
Rover others                1
Fiat suv                    1
Chrysler suv                1
Land_rover convertible      1
Lada convertible            1
Volvo small car             1
Saab others                 1
Lancia coupe                1
Subaru others               1
Suzuki coupe                1
Skoda convertible           1
Smart station wagon         1
Lada unknown                1
dtype: int64

**Average price by vehicle type**

In [ ]:
price_type = autos_selected[["vehicle_type","price"]] 

In [ ]:
price_type["vehicle_type"].value_counts().sum()
mean_price_type = {}

for vehicle in price_type["vehicle_type"].unique():
    vehicle_ = price_type[price_type["vehicle_type"] == vehicle]
    type_mean = vehicle_["price"].mean()
    mean_price_type[vehicle] = round(type_mean,2)

mean_price_top_type = []
    
for k,v in mean_price_type.items():
    key_val_as_tuple = (mean_price_type[k], k)
    mean_price_top_type.append(key_val_as_tuple)

mean_price_top_type = sorted(mean_price_top_type, reverse = True)

vehicle_type = []
price_ = []

for vehicle in mean_price_top_type:
    vehicle_type.append(vehicle[1].capitalize())
    price_.append(vehicle[0])

top_mean_prices_vehicle_type = pd.DataFrame(price_,index = vehicle_type,columns = ["mean_price"])

In [ ]:
top_mean_prices_vehicle_type

mean_price
Suv               7804.64
Convertible       6029.57
Coupe             5180.17
Bus               5054.19
Station wagon     4535.59
Limousine         4389.17
Others            3594.67
Small car         2619.65
Unknown           2028.37

**How much cheaper are cars with damage than their non-damaged counterparts**

In [ ]:
unrepaired_damage = round(autos_selected[autos_selected["unrepaired_damage"] == "no"]["price"].mean(),2)
repaired_damage = round(autos_selected[autos_selected["unrepaired_damage"] == "yes"]["price"].mean(),2)
unknown_repaired = round(autos_selected[autos_selected["unrepaired_damage"] == "unknown"]["price"].mean(),2)

print("Average price for used vehicles with unrepaired damage: €",unrepaired_damage)
print("Average price for used vehicles with repaired damage: €",repaired_damage)
print("Average price for used vehicles with unknown repaired damage: €",unknown_repaired)

Average price for used vehicles with unrepaired damage: € 4851.27
Average price for used vehicles with repaired damage: € 1972.5
Average price for used vehicles with unknown repaired damage: € 2619.24


**At the end of the cleanup, I store the processed data frame in .csv format as this could do further analysis without requiring the cleanup process again.**

In [ ]:
autos_selected["power_PS"] = autos_selected["power_PS"].replace(0,"unknown")
autos_selected.to_csv('autos_clean.csv', index=False)

**I encountered a certain error with the names, visible in the .csv format, this is that some names have characters "," or ";" (delimiters) that when exporting in .csv format make their corresponding information move between columns. So I decided to replace those characters in the names with underscores.**

In [ ]:
autos_selected["name"] = autos_selected["name"].replace(";","_").replace(",","_")
autos_selected.to_csv('autos_clean.csv', index=False)

In [ ]:
autos_selected

date_crawled  ...            last_seen
0      2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1      2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2      2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3      2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4      2016-04-01 14:38:50  ...  2016-04-01 14:38:50
...                    ...  ...                  ...
49993  2016-03-15 18:47:35  ...  2016-04-06 19:46:53
49994  2016-03-22 17:36:42  ...  2016-04-06 08:16:39
49996  2016-03-28 10:50:25  ...  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  ...  2016-04-04 11:47:27
49999  2016-03-14 00:42:12  ...  2016-04-06 21:18:48

[42959 rows x 17 columns]

# Conclusion

- Used cars for sale with registration year between 1910 and 2016, the highest frequency corresponds to those with registration year 2000 (approximately 7,3%). Most used cars for sale (51%) have their registration years between 1999 and 2006.


- The maximum average selling price by brand corresponds to Porsche brand vehicles with an average value of € 9,493.25. The minimum average selling price by brand corresponds to Daewoo brand vehicles with an average value of € 1,049.00.


- The highest mileage used cars are Saab brand with 145,526.32 km average. The lowest-mileage used cars are from the Trabant brand with an average of 54,538.46 km


- The highest power used cars are Porsche brand with 246.25 PS average. The lowest power used cars are Trabant brand with 27.65 PS average.


- The brands by most common model is Volkswagen golf with 3487 cars. The Less common brands by model are Volvo v60, Land Rover others, Audi 200, Bmw i3, Rover discovery, Jaguar unknown, Land_rover unknown, Land_rover range_rover_evoque and Ford b_max.


- The most common brand by vehicle type are Volkswagen limousines. The least common brands by vehicle type are Saab others, Land_rover convertible, Smart station wagon, Land_rover station wagon, Subaru others, Skoda convertible, Suzuki coupe, Saab suv, Lancia coupe, Lada small car, Volvo small car, Chrysler suv, Smart others, Lada unknown, Rover others, Lada convertible, Jeep convertible, Fiat suv and Skoda coupe.


- The highest average price per vehicle type corresponds to "Sub" vehicle type with € 7,804.64. The lowest average price per vehicle type corresponds to Unknown Car type vehicles with € 2,028.37.


- Vehicles with unrepaired damage are sold on average at a price 2.5 higher than the price of vehicles with repaired damage.Vehicles with unknown damage or repairs have an average price of € 2,619.24.